### ISSUE TITLE
failFast & --enable-default-rulesets confusion

### ISSUE THREAD

**Issue by 3flex** [arturbosch/detekt#663]
To me the meaning of "failFast" in the config file is unclear. In other contexts it usually means "fail as soon as a violation is found", or at least that's how it's used in Travis and Rubocop. Having that option would speed up tests where required.

Another confusing thing is the CLI has the `--disable-default-rulesets` option, which is basically the opposite of "failFast", so to enable all rules, you have to say "failFast" in the config file, but to disable all rules, you have to say `--disable-default-rulesets` on the command line.

It would be better to have enabling all rule and disabling default rules available in the same place (the options could be set in the config file, then overridden on the CLI).

**Comment by arturbosch** [comment]
`--disable-default-rulesets` was introduced when formatting and migration modules could be used by the gradle plugin. This flag was a easy way just to run the formatting rules etc.
It still has the purpose if the user just want his own rules to run.

I agree with you. Having `failfast` and `disable-default-rulesets` in the same place is desirable.

I propose to just have both options as flags inside the cli and corresponding gradle-plugin tasks.
What do you think? @3flex @Mauin @schalkms @vanniktech

**Comment by vanniktech** [comment]
I'm against removing the failFast property from the configuration. I don't see the benefit of having multiple Gradle tasks. I'd just push the other property in the configuration and remove it from the CLI.

**Comment by 3flex** [comment]
To keep everyone happy it would go in both places. If it's going to be in one place only, I agree that it would go in the CLI.

I'm not sure the argument against multiple Gradle tasks holds - you'd either have to specify the CLI flag, or a different YAML config either way right? Unless I missed something:
```gradle
detekt {
    profile("failFast") {
        config = "$projectDir/failFast.yml"
   }
    profile("disableByDefault") {
        config = "$projectDir/disableByDefault.yml"
   }   
}
```
vs
```gradle
detekt {
    profile("failFast") {
        failFast = true // passed to CLI
   }
    profile("disableByDefault") {
        disableDefaultRuleSets = true // passed to CLI
   }   
}
```

**Comment by lwasylkowski** [comment]
I was also confused by the `failFast` name, but I'm strongly against removing it from the configuration file, as it's strictly related to the configuration itself. Changing it to something like `rulesDefaultState: [active|inactive|default]` is way better. I'm not sure about the name, but configuration should be contained in one place. Actually I don't see why would anyone have a config file which they would want to run with both `failFast` and `disableByDefault` options

**Comment by vanniktech** [comment]
We could add it in both places, yes.

Maybe failFast just needs better documentation since the naming does not quite convey the intention.

What's the advantage of having multiple profiles? I never really had the need for that.

**Comment by 3flex** [comment]
@arturbosch I'd originally raised two issues here: one was that it was inconsistent that `--disable-default-rulesets` could be set on CLI but fail fast option couldn't (this was fixed in https://github.com/arturbosch/detekt/pull/1417).

The other was the use of the term "fail fast" which is still different to all other usages I've come across for lint/testing tools, but seems here to stay.

I think this can be closed unless you think there's something else that needs discussing here?
